File: 2025_Oct/evelyn/week6/2025-11-24-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> inorderMap;
    int pRootIdx = 0;

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = inorder.size();
        for (int i = 0; i < n; i++) inorderMap[inorder[i]] = i;

        return helper(preorder, 0, n - 1);
    }

    TreeNode* helper(vector<int>& preorder, int left, int right) {
        if (left > right) return nullptr;

        int rootVal = preorder[pRootIdx];
        pRootIdx++;

        TreeNode* root = new TreeNode(rootVal);
        int iRootIdx = inorderMap[rootVal];

        root->left = helper(preorder, left, iRootIdx - 1);
        root->right = helper(preorder, iRootIdx + 1, right);

        return root;
    }
};

File: 2025_Oct/Jayden/2025-11-24-Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
        Time Complexity: O(n^2)
            - finding root index in inorder takes O(n) for each recursive call
        Space Complexity: O(1)

        1) The first value of the preorder array is the root of the tree.
        2) The second value of the preorder array is the root of the left sub tree
        3) All values on the left side of the inorder from root are the nodes of the left subtree
        4) All values on the right side of the inorder from root are the nodes of the right subtree
     */
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0 || inorder.length == 0) return null;

        TreeNode root = new TreeNode(preorder[0]);
        Integer mid = null;

        for (int i = 0; i < inorder.length; i++) {
            if (root.val == inorder[i]) mid = i;
        }

        if (mid == null) return null;

        root.left = buildTree(Arrays.copyOfRange(preorder, 1, mid + 1), Arrays.copyOfRange(inorder, 0, mid));
        root.right = buildTree(Arrays.copyOfRange(preorder, mid + 1, preorder.length), Arrays.copyOfRange(inorder, mid + 1, inorder.length));

        return root;
    }
}